policy/modules/services/incus.te

```diff
@@ -0,0 +1,198 @@
+
+# for bridge management
+dev_create_sysfs_files(incusd_t)
+dev_write_sysfs(incusd_t)
+
+# incusd wont start without access to /dev/urandom
+dev_read_urand(incusd_t)
+
+# mounton access to /proc/sys/kernel/random/boot_id
+kernel_mounton_kernel_sysctl_files(incusd_t)
+
+# read /etc/machine-id
+files_read_etc_runtime_files(incusd_t)
+
+# watch /dev/hugepages
+files_watch_all_dirs(incusd_t)
+
+# incus apparmor support wants to handle /sys/kernel/tracing
+fs_dontaudit_getattr_tracefs(incusd_t)
+fs_dontaudit_getattr_tracefs_dirs(incusd_t)
+fs_dontaudit_remount_tracefs(incusd_t)
+
+# reason for access not (yet?) looked up in the source code
+# but everything seems to work fine so far
+fs_dontaudit_getattr_configfs(incusd_t)
+fs_dontaudit_remount_configfs(incusd_t)
+
+# required for fanotify on /dev/zvol/zdata/incus/virtual-machines/.#vm1.block...
+dev_watch_dev_fs(incusd_t)
+
+# required so that incus will get the VM PID
+# and make 'incus info VM' work
+# Error: stat /proc/628420: permission denied
+qemu_read_state(incusd_t)
+
+# TODO: would it be good to have some "legacy_mounts" bool and let it allow to do
+# with this mounts what incus/lxc wants it to do?
+#
+# from doc/container-environment.md:
+#
+# If they are present on the host, the following paths will also automatically be mounted:
+#
+# - `/proc/sys/fs/binfmt_misc`
+# - `/sys/firmware/efi/efivars`
+# - `/sys/fs/fuse/connections`
+# - `/sys/fs/pstore`
+# - `/sys/kernel/debug`
+# - `/sys/kernel/security`
+#
+# The reason for passing all of those paths is that legacy init systems require
+# them to be mounted, or be mountable, inside the container.
+#
+# The majority of those paths will not be writable (or even readable) from
+# inside an unprivileged container. In privileged containers, they will be
+# blocked by the AppArmor policy.
+#
+fs_dontaudit_getattr_binfmt_misc_fs(incusd_t)
+
+fs_dontaudit_getattr_efivarfs(incusd_t)
+fs_dontaudit_remount_efivarfs(incusd_t)
+
+fs_dontaudit_getattr_pstorefs(incusd_t)
+fs_dontaudit_remount_pstorefs(incusd_t)
+
+kernel_dontaudit_search_debugfs(incusd_t)
+fs_dontaudit_getattr_debugfs(incusd_t)
+fs_dontaudit_remount_debugfs(incusd_t)
+
+# incus idmap support: checks whether the kernel supports VFS v3 fscaps
+container_execute_engine_tmp_files(incusd_t)
+
+dontaudit incusd_t self:capability sys_module;
+
+# required for QEMU feature check, won't start otherwise
+files_rw_generic_tmp_sockets(incusd_t)
+files_dontaudit_unlink_generic_tmp_sockets(incusd_t)
+
+qemu_stream_connect(incusd_t)
+qemu_kill(incusd_t)
+dev_getattr_vhost_dev(incusd_t)
+
+# these are required for incusd to check and execute qemu
+qemu_domtrans(incusd_t)
+qemu_exec(incusd_t)
+
+# incusd needs support to label VM files before launching
+# until then, VM files will unfortunately be unlabeled
+kernel_manage_unlabeled_dirs(incusd_t)
+kernel_manage_unlabeled_files(incusd_t)
+kernel_manage_unlabeled_symlinks(incusd_t)
+kernel_mounton_unlabeled_dirs(incusd_t)
+files_search_default(incusd_t)
```

Comment:
Does this fuctionality currently exist in incus?

Reply:
Incus is still missing some selinux awareness, e.g. label some objects properly before launching a qemu VM. Until it will get support for that we can remove these `kernel_manage_*` and (I guess) `files_search_default` again.

Does this answer your question?

Some more details:
```
# audit(1734698529.636:163140):
#  scontext="system_u:system_r:incusd_t:s0" tcontext="system_u:object_r:unlabeled_t:s0"
#  class="dir" perms="getattr"
#  comm="incusd" exe="" path=""
#  message="type=AVC msg=audit(1734698529.636:163140): avc:  denied  { getattr }
#   for  pid=1166765 comm="incusd" path="/var/lib/incus/storage-
#   pools/default/virtual-machines/vm1" dev="zfs" ino=34
#   scontext=system_u:system_r:incusd_t:s0
#   tcontext=system_u:object_r:unlabeled_t:s0 tclass=dir permissive=0 "
allow incusd_t unlabeled_t:dir getattr;

(sysadm_r)@host ~ # mount|grep vm1
zdata/incus/virtual-machines/vm1 on /var/lib/incus/storage-pools/default/virtual-machines/vm1 type zfs (rw,relatime,seclabel,xattr,posixacl,casesensitive)
zdata/incus/virtual-machines/vm1 on /var/lib/incus/devices/vm1/config.mount type zfs (ro,relatime,seclabel,xattr,posixacl,casesensitive)

(sysadm_r)@host ~ # ls -lZR /var/lib/incus/virtual-machines
/var/lib/incus/virtual-machines:
total 9
lrwxrwxrwx. 1 root root system_u:object_r:container_var_lib_t:s0 57 20. Dez 13:44 vm1 -> /var/lib/incus/storage-pools/default/virtual-machines/vm1
(sysadm_r)@host ~/selinux/local # ls -lZRd /var/lib/incus/virtual-machines/vm1
lrwxrwxrwx. 1 root root system_u:object_r:container_var_lib_t:s0 57 20. Dez 13:44 /var/lib/incus/virtual-machines/vm1 -> /var/lib/incus/storage-pools/default/virtual-machines/vm1
(sysadm_r)@host ~ # 
(sysadm_r)@host ~ # ls -lZR /var/lib/incus/storage-pools/default/virtual-machines/vm1
/var/lib/incus/storage-pools/default/virtual-machines/vm1:
ls: cannot read symbolic link '/var/lib/incus/storage-pools/default/virtual-machines/vm1/qemu.nvram': Permission denied
total 73
-rw-r--r--. 1 root   root system_u:object_r:unlabeled_t:s0    688 20. Dez 13:44 agent-client.crt
-rw-------. 1 root   root system_u:object_r:unlabeled_t:s0    288 20. Dez 13:44 agent-client.key
-rw-r--r--. 1 root   root system_u:object_r:unlabeled_t:s0    725 20. Dez 13:44 agent.crt
-rw-------. 1 root   root system_u:object_r:unlabeled_t:s0    288 20. Dez 13:44 agent.key
-r--------. 1 root   root system_u:object_r:unlabeled_t:s0   2762 20. Dez 13:44 backup.yaml
dr-x------. 6 nobody root system_u:object_r:unlabeled_t:s0     14 20. Dez 13:44 config
-rw-------. 1 nobody root system_u:object_r:unlabeled_t:s0 540672 20. Dez 13:44 edk2-i386-vars.fd
-rw-r--r--. 1 root   root system_u:object_r:unlabeled_t:s0    526 20. Dez 08:48 metadata.yaml
lrwxrwxrwx. 1 root   root system_u:object_r:unlabeled_t:s0     17 20. Dez 13:44 qemu.nvram
drwxr-xr-x. 2 root   root system_u:object_r:unlabeled_t:s0      4 20. Dez 08:48 templates

total 13718
-r--------. 1 nobody root system_u:object_r:unlabeled_t:s0        2 20. Dez 13:44 agent-mounts.json
-rw-r--r--. 1 nobody root system_u:object_r:unlabeled_t:s0      818 20. Dez 13:44 agent.conf
-r--------. 1 nobody root system_u:object_r:unlabeled_t:s0      725 20. Dez 13:44 agent.crt
-r--------. 1 nobody root system_u:object_r:unlabeled_t:s0      288 20. Dez 13:44 agent.key
dr-x------. 2 nobody root system_u:object_r:unlabeled_t:s0        3 20. Dez 13:44 files
-r-x------. 1 nobody root system_u:object_r:unlabeled_t:s0 21724049 12. Dez 17:08 incus-agent
-rwx------. 1 nobody root system_u:object_r:unlabeled_t:s0     1302 20. Dez 13:44 install.sh
lrwxrwxrwx. 1 root   root system_u:object_r:unlabeled_t:s0       11 20. Dez 13:44 lxd-agent
dr-x------. 2 nobody root system_u:object_r:unlabeled_t:s0        2 20. Dez 13:44 nics
-r--------. 1 nobody root system_u:object_r:unlabeled_t:s0      688 20. Dez 13:44 server.crt
dr-x------. 2 nobody root system_u:object_r:unlabeled_t:s0        4 20. Dez 13:44 systemd
dr-x------. 2 nobody root system_u:object_r:unlabeled_t:s0        3 20. Dez 13:44 udev

/var/lib/incus/storage-pools/default/virtual-machines/vm1/config/files:
total 5
-rw-r--r--. 1 nobody root system_u:object_r:unlabeled_t:s0 526 20. Dez 13:44 metadata.yaml

/var/lib/incus/storage-pools/default/virtual-machines/vm1/config/nics:
total 0

/var/lib/incus/storage-pools/default/virtual-machines/vm1/config/systemd:
total 9
-r-x------. 1 nobody root system_u:object_r:unlabeled_t:s0 1570 20. Dez 13:44 incus-agent-setup
-r--------. 1 nobody root system_u:object_r:unlabeled_t:s0  419 20. Dez 13:44 incus-agent.service

/var/lib/incus/storage-pools/default/virtual-machines/vm1/config/udev:
total 5
-r--------. 1 nobody root system_u:object_r:unlabeled_t:s0 109 20. Dez 13:44 99-incus-agent.rules

/var/lib/incus/storage-pools/default/virtual-machines/vm1/templates:
total 9
-rw-r--r--. 1 root root system_u:object_r:unlabeled_t:s0  21 20. Dez 08:48 hostname.tpl
-rw-r--r--. 1 root root system_u:object_r:unlabeled_t:s0 140 20. Dez 08:48 hosts.tpl


```